source/common/common/fancy_logger.cc

```diff
@@ -0,0 +1,95 @@
+#include "common/common/fancy_logger.h"
+
+#include <atomic>
+#include <memory>
+
+#include "common/common/logger.h"
+
+using spdlog::level::level_enum;
+
+namespace Envoy {
+
+absl::Mutex FancyContext::fancy_log_lock_(absl::kConstInit);
+
+FancyMapPtr FancyContext::fancy_log_map_ = std::make_shared<FancyMap>();
+
+/**
+ * Implements a lock from BasicLockable, to avoid dependency problem of thread.h.
+ */
+class FancyBasicLockable : public Thread::BasicLockable {
+public:
+  // BasicLockable
+  void lock() ABSL_EXCLUSIVE_LOCK_FUNCTION() override { mutex_.Lock(); }
+  bool tryLock() ABSL_EXCLUSIVE_TRYLOCK_FUNCTION(true) override { return mutex_.TryLock(); }
+  void unlock() ABSL_UNLOCK_FUNCTION() override { mutex_.Unlock(); }
+
+private:
+  absl::Mutex mutex_;
+};
+
+SpdLoggerPtr FancyContext::getFancyLogEntry(std::string key) ABSL_LOCKS_EXCLUDED(fancy_log_lock_) {
+  absl::ReaderMutexLock l(&fancy_log_lock_);
+  return fancy_log_map_->find(key)->second;
+}
+
+void FancyContext::initFancyLogger(std::string key, std::atomic<spdlog::logger*>& logger)
+    ABSL_LOCKS_EXCLUDED(fancy_log_lock_) {
+  absl::WriterMutexLock l(&FancyContext::fancy_log_lock_);
+  auto it = fancy_log_map_->find(key);
+  spdlog::logger* target;
+  if (it == fancy_log_map_->end()) {
+    target = createLogger(key);
+  } else {
+    target = it->second.get();
+  }
+  logger.store(target);
+}
+
+bool FancyContext::setFancyLogger(std::string key, level_enum log_level)
+    ABSL_LOCKS_EXCLUDED(fancy_log_lock_) {
+  absl::ReaderMutexLock l(&FancyContext::fancy_log_lock_);
+  auto it = fancy_log_map_->find(key);
+  if (it != fancy_log_map_->end()) {
+    it->second->set_level(log_level);
+    return true;
+  }
+  return false;
+}
+
+void FancyContext::setDefaultFancyLevelFormat(spdlog::level::level_enum level, std::string format)
```

Comment:
I think you want to compare the previous default level vs the new default level. So that you can avoid the for loop if nothing changes.

Also, you can compare the logger's log level vs default log level so that you can avoid overriding the log level set to non default value.

Reply:
Sure.